AndroidSDKCore/src/main/java/com/leanplum/internal/Request.java

```diff
@@ -480,22 +480,88 @@ protected Void doInBackground(Void... params) {
     });
   }
 
-  private void sendRequests() {
+
+  /**
+   * This class wraps the unsent requests, requests that we need to send
+   * and the JSON encoded string. Wrapping it in the class allows us to
+   * retain consistency in the requests we are sending and the actual
+   * JSON string.
+   */
+  static class RequestsWithEncoding {
+    List<Map<String, Object>> unsentRequests;
+    List<Map<String, Object>> requestsToSend;
+    String jsonEncodedString;
+  }
+
+  private RequestsWithEncoding getRequestsWithEncodedStringForErrors() {
     List<Map<String, Object>> unsentRequests = new ArrayList<>();
     List<Map<String, Object>> requestsToSend;
+    String jsonEncodedRequestsToSend;
+
+    String uuid = UUID.randomUUID().toString();
+    for (Map<String, Object> error : localErrors) {
+      error.put(UUID_KEY, uuid);
+      unsentRequests.add(error);
+    }
+    requestsToSend = unsentRequests;
+    jsonEncodedRequestsToSend = jsonEncodeUnsentRequests(unsentRequests);
+
+    RequestsWithEncoding requestsWithEncoding = new RequestsWithEncoding();
+    // for errors, we send all unsent requests so they are identical
+    requestsWithEncoding.unsentRequests = unsentRequests;
+    requestsWithEncoding.requestsToSend = requestsToSend;
+    requestsWithEncoding.jsonEncodedString = jsonEncodedRequestsToSend;
+
+    return requestsWithEncoding;
+  }
+
+
+  protected RequestsWithEncoding getRequestsWithEncodedStringStoredRequests(double fraction) {
+    try {
+      List<Map<String, Object>> unsentRequests;
+      List<Map<String, Object>> requestsToSend;
+      String jsonEncodedRequestsToSend;
+      RequestsWithEncoding requestsWithEncoding = new RequestsWithEncoding();
+
+      if (fraction < 0.01) { //base case
```

Comment:
@e7mac Is there a reason, we are dropping the request after ~6 attempts i.e around 100 requests?
Could we add some server-side logging or at least a metric here so we know when this happens?

Reply:
@e7mac Also, when it hits this code, it means nothing succeeded so far. Shouldn't we clear the entire existing requests in `LeanplumEventDataManager.getEvents` otherwise there will be another attempt to send all these requests the next time?

Reply:
@divyas1102 We have server side logging and we will be monitoring that closely as we roll out the long-term fix.
We are looking for a solution without data loss. The current behavior of crashing is unacceptable and is being addressed here. Beyond that, we want to preserve the data and try sending it later when network and/or memory conditions might be better.